### PR TITLE
Spring hw 11 solution

### PR DESCRIPTION
**Pull request by AnatolyNz**

### CODE REVIEW

File: src/main/java/mate/academy/dto/CreateBookRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+public class CreateBookRequestDto {
+    @NotNull
+    private String title;
+    @NotNull
+    private String author;
+    private String isbn;
+    @NotNull
+    @Min(0)
+    private BigDecimal price;
+    private String description;
```
Comment (MentorMA) [inline src/main/java/mate/academy/dto/CreateBookRequestDto.java]: Lets make better validation, use more direct anno like @notblank @isbn, etc

File: src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                .toList();
+        body.put("errors", errors);
+        return new ResponseEntity<>(body, headers, status);
+    }
+
+    private String getErrorMessage(ObjectError e) {
+        if (e instanceof FieldError) {
+            String field = ((FieldError) e).getField();
+            String message = e.getDefaultMessage();
+            return field + SPACE + message;
```
Comment (MentorMA) [inline src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java:49]:
```suggestion
        if (e instanceof FieldError field) {
            String message = e.getDefaultMessage();
            return field + SPACE + message;
```

Comment (MentorMA) [reply src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java:49]: Try to use new feature of java

File: src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import org.springframework.validation.ObjectError;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@ControllerAdvice
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    private static final String SPACE = " ";
```
Comment (MentorMA) [inline src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java]: Try to find constant of empty string StringUtils.EMPTY or smth else

File: src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            HttpStatusCode status,
+            WebRequest request
+    ) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
+        List<String> errors = ex.getBindingResult().getAllErrors().stream()
+                .map(this::getErrorMessage)
+                .toList();
+        body.put("errors", errors);
```
Comment (MentorMA) [inline src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java]: Let`s save strings to constants

File: src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Object> handleRegistrationException(
+            RegistrationException ex,
+            WebRequest request) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
+        body.put("message", ex.getMessage());
```
Comment (MentorMA) [inline src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java]: here you can reuse constants

File: src/main/java/mate/academy/repository/BookDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Override
+    public Optional<Book> getBookById(Long id) {
+        try (Session session = sessionFactory.openSession()) {
+            return session.createQuery("SELECT book FROM Book book WHERE book.id = :id", Book.class)
+                    .setParameter("id", id)
+                    .uniqueResultOptional();
+        } catch (Exception e) {
+            throw new RuntimeException("Can't get book by id from DB: " + id, e);
```
Comment (MentorMA) [inline src/main/java/mate/academy/repository/BookDaoImpl.java]: You have EntityNotFoundException

File: src/main/java/mate/academy/repository/book/BookSpecificationBuilder.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+                    .getSpecification(searchParameters.titles()));
+        }
+        if (searchParameters.authors() != null && searchParameters.authors().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("author")
+                    .getSpecification(searchParameters.authors()));
+        }
+        if (searchParameters.prices() != null && searchParameters.prices().length > 0) {
+            spec = spec.and(bookSpecificationProviderManager
+                    .getSpecificationProvider("price")
```
Comment (MentorMA) [inline src/main/java/mate/academy/repository/book/BookSpecificationBuilder.java]: Save string to constants

File: src/main/java/mate/academy/repository/book/spec/AuthorSpecificationProvider.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+import mate.academy.model.Book;
+import mate.academy.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public String getKey() {
+        return "author";
```
Comment (MentorMA) [inline src/main/java/mate/academy/repository/book/spec/AuthorSpecificationProvider.java]: Or save constat here and reuse it where it need

**Review by MentorMA — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java (hunk 2 above)
Comment (AnatolyNz) [reply src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java:49]: I can't do so, I receive field in return.

File: src/main/java/mate/academy/dto/CreateBookRequestDto.java [hunk 9]
```diff
@@ -1,18 +1,18 @@
...
+    @NotBlank
     private String title;
-    @NotNull
+    @NotBlank
     private String author;
     private String isbn;
-    @NotNull
+    @NotBlank
     @Min(0)
     private BigDecimal price;
```
Comment (MentorMA) [inline src/main/java/mate/academy/dto/CreateBookRequestDto.java]: Validate isbn

File: src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java [hunk 10]
```diff
@@ -18,7 +19,11 @@
 
 @ControllerAdvice
 public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
-    private static final String SPACE = " ";
+    private static final String SPACE = StringUtils.EMPTY;
+    private static final String ERRORS_PARAM = "errors";
+    private static final String MESSAGE_PARAM = "message";
+    private static final String STATUS_PARAM = "status";
+    private static final String TIMESTAMP_PARAM = "timestamp";
```
Comment (MentorMA) [inline src/main/java/mate/academy/exception/CustomGlobalExceptionHandler.java]: as you wish you can use naming with out param

File: src/main/java/mate/academy/repository/book/BookSpecificationBuilder.java [hunk 11]
```diff
@@ -12,24 +12,27 @@
 @RequiredArgsConstructor
 @Component
 public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private static final String AUTHOR_PARAM = "author";
+    private static final String PRICE_PARAM = "price";
+    private static final String TITLE_PARAM = "title";
```
Comment (MentorMA) [inline src/main/java/mate/academy/repository/book/BookSpecificationBuilder.java]: The same

File: src/main/java/mate/academy/repository/book/spec/AuthorSpecificationProvider.java [hunk 12]
```diff
@@ -8,13 +8,15 @@
 
 @Component
 public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    private static final String AUTHOR_PARAM = "author";
```
Comment (MentorMA) [inline src/main/java/mate/academy/repository/book/spec/AuthorSpecificationProvider.java]: You already have constant

**Review by MentorMA — CHANGES_REQUESTED** [review]

**Review by kozhukhovsky — APPROVED** [review]

**Pull request merged**